### DISCUSSION TITLE
Invalid prisma.$queryRawUnsafe() invocation

### DISCUSSION THREAD

**Discussion by thammer67** [prisma/prisma#24495] [Q&A] [answered]
### Question

I am writing an app in Nuxt 3.11 using Prisma 5.15.0 and @prisma/client 5.15.0.

I have two api endpoints using prisma.$queryRawUnsafe(). One works fine but the other produces the following error:

```
ERROR  [nuxt] [request error] [unhandled] [500] 
Invalid prisma.$queryRawUnsafe() invocation:

Raw query failed. Code: N/A. Message: N/A
  at In.handleRequestError (./node_modules/@prisma/client/runtime/library.js:122:6877)  
  at In.handleAndLogRequestError (./node_modules/@prisma/client/runtime/library.js:122:6211)  
  at In.request (./node_modules/@prisma/client/runtime/library.js:122:5919)  
  at async l (./node_modules/@prisma/client/runtime/library.js:127:11167)  
  at Object.handler (./server/api/search/Clients.js:9:1)  
  at async ./node_modules/h3/dist/index.mjs:1962:19  
  at async Object.callAsync (./node_modules/unctx/dist/index.mjs:72:16)  
  at async Server.toNodeHandle (./node_modules/h3/dist/index.mjs:2249:7)
```

The API endpoint that works is as follows:
```js
export default defineEventHandler(async event=> {
    const body = await readBody(event)

    const res = await prisma.$queryRawUnsafe(`
        SELECT * FROM users 
        WHERE
            firstname LIKE ? 
            OR lastname LIKE ? 
            OR email LIKE ? 
            OR phone1 LIKE ?
            OR phone2 LIKE ?
    `, `%${body.text}%`, `%${body.text}%`, `%${body.text}%`, `%${body.text}%`, `%${body.text}%`)

    return res
})
```

The API endpoint that produces the error is nearly identical except for a different table:
```js
export default defineEventHandler(async event=> {

    const body = await readBody(event)

    const res = await prisma.$queryRawUnsafe(`
        SELECT * FROM clients
        WHERE
            name LIKE ? 
            OR email LIKE ? 
            OR phone1 LIKE ?
    `, `%${body.text}%`, `%${body.text}%`, `%${body.text}%`)

    return res
})
```

I have defined the prisma object in server/utils/prisma.js as follows:
```js
import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()

export default prisma
```

Please help. I cannot figure out what I am doing wrong here. 

### Information about Prisma Schema, Client Queries and Environment (optional)
```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mysql"
  url      = env("DATABASE_URL")
}

model clients {
  serial      Int       @id @default(autoincrement())
  name        String    @default("") @db.VarChar(255)
  phone1      String    @default("") @db.VarChar(30)
  phone2      String    @default("") @db.VarChar(30)
  email       String    @default("") @db.VarChar(255)

  @@index([name], map: "name")
}

model users {
  serial         Int                   @id @default(autoincrement())
  firstname      String                @default("") @db.VarChar(255)
  lastname       String                @default("") @db.VarChar(255)
  email          String                @default("") @db.VarChar(255)
  phone1         String                @default("") @db.VarChar(30)
  phone2         String                @default("") @db.VarChar(30)

  @@index([email], map: "email")
  @@index([lastname], map: "lastname")
}
```

**Comment by nurul3101** [comment]
Hello @thammer67 👋 

Can you try running the SQL statement directly against your database and check if you get the same error?
```sql
SELECT * FROM clients
        WHERE
            name LIKE ? 
            OR email LIKE ? 
            OR phone1 LIKE ?
    `, `%${body.text}%`, `%${body.text}%`, `%${body.text}%`
```

Also, try adding DEBUG="*" in your env variables as recommended [here](https://www.prisma.io/docs/orm/prisma-client/debugging-and-troubleshooting/debugging#setting-the-debug-environment-variable) and check if you get some more error information.

**Reply by thammer67** [reply]
Thanks for your suggestions!

Yes, I can run the SQL query directly against the database with no problems. I tried removing the other (working) API endpoint so that I only have the one endpoint and query and the problem still happens. So then I tried to simplify the clients query to see if I could get it to succeed like so:
```
select count(*) from clients
```
That seems to work fine with no errors. So I tried to expand the query slightly:
```
select * from clients LIMIT 10
```
and that produced the error.

Note that if I change the file to query the users table (select * from users LIMIT 10) there is no error, which leads me to think Prisma has a problem understanding the clients table as defined. So I re-ran 'npx prisma generate' to be sure and the problem still happens.

I will try setting the DEBUG option and see what I get.

**Reply by thammer67** [reply]
Also, I also tried another query and it succeeds without errors:
```
SELECT 
            FORMAT(count(*),0) 'Total Clients', 
            FORMAT(SUM(CASE WHEN active = 1 THEN 1 ELSE 0 END),0) 'Active Clients', 
            FORMAT(SUM(CASE WHEN active = 0 THEN 1 ELSE 0 END),0) 'Inactive Clients' 
        FROM clients
```
So I went back to the query:
```
SELECT * FROM clients LIMIT 10
```

Here is the debug output as you requested. I don't see anything useful there though:
```
[nitro-runtime] request: 0.63ms
prisma:client prisma.$queryRaw(
            select * from clients limit 10
        , []) +8259ms
Server middleware (authValidate), to: [POST] /api/search/Clients, from: undefined
/api/Clients, search text:  asdfsf
prisma:client Prisma Client call: +0ms
prisma:client prisma.$queryRaw({
  query: "\n            select * from clients limit 10\n        ",
  parameters: {
    values: "[]",
    __prismaRawParameters__: true
  }
}) +1ms
prisma:client Generated request: +0ms
prisma:client {
  "action": "queryRaw",
  "query": {
    "arguments": {
      "query": "\n            select * from clients limit 10\n        ",
      "parameters": "[]"
    },
    "selection": {}
  }
}
 +0ms
prisma:client:libraryEngine sending request, this.libraryStarted: false +1ms
prisma:client:libraryEngine library starting +0ms
prisma:client:libraryEngine library started +3ms
prisma:client:request_handler {
  "name": "PrismaClientKnownRequestError",
  "code": "P2010",
  "clientVersion": "5.15.0",
  "meta": {
    "code": "N/A",
    "message": "N/A"
  }
} +2ms

Invalid prisma.$queryRaw() invocation:


Raw query failed. Code: N/A. Message: N/A

  Invalid prisma.$queryRaw() invocation:
  
  
  Raw query failed. Code: N/A. Message: N/A
  at In.handleRequestError (node_modules/@prisma/client/runtime/library.js:122:6877)
  at In.handleAndLogRequestError (node_modules/@prisma/client/runtime/library.js:122:6211)
  at In.request (node_modules/@prisma/client/runtime/library.js:122:5919)
  at async l (node_modules/@prisma/client/runtime/library.js:127:11167)
  at Object.handler (server/api/search/Clients.js:10:1)
  at async node_modules/h3/dist/index.mjs:1962:19
  at async Object.callAsync (node_modules/unctx/dist/index.mjs:72:16)
  at async Server.toNodeHandle (node_modules/h3/dist/index.mjs:2249:7)
[nitro-runtime] beforeResponse: 0.03ms
[nitro-runtime] dev:ssr-logs: 0.033ms
[nitro-runtime] afterResponse: 1.087ms
```

**Reply by thammer67** [reply]
I have been "simplifying" my schema to make it easier to debug. Here is the actual schema of my "clients" table. I am wondering if there is something in the way Prisma introspected the existing table that is causing my errors. 
```
/// This model or at least one of its fields has comments in the database, and requires an additional setup for migrations: Read more: https://pris.ly/d/database-comments
model c_clients {
  c_serial      Int       @id @default(autoincrement())
  c_name        String    @default("") @db.VarChar(255)
  c_ic_serial   Int       @default(0)
  c_address1    String    @default("") @db.VarChar(255)
  c_address2    String    @default("") @db.VarChar(255)
  c_city        String    @default("") @db.VarChar(255)
  c_state       String    @default("") @db.Char(2)
  c_zip         String    @default("") @db.VarChar(5)
  c_phone1      String    @default("") @db.VarChar(30)
  c_phone2      String    @default("") @db.VarChar(30)
  c_fax         String    @default("") @db.VarChar(30)
  c_email       String    @default("") @db.VarChar(255)
  c_clientsince DateTime  @default(dbgenerated("'0000-00-00'")) @db.Date
  c_dtec        DateTime? @db.DateTime(0)
  c_dtem        DateTime? @db.DateTime(0)
  c_active      Boolean   @default(true)
  c_createdby   Int       @default(0)
  c_modifiedby  Int       @default(0)

  @@index([c_name], map: "c_name")
}
```

**Reply by thammer67** [reply]
So I think there is problem is with the `c_clientsince` column in my `c_clients` table. I have been testing different queries and I am not getting any errors with any queries where I select any column name except `c_clientsince`. When I select `c_clientsince`, the N/A error is thrown. At least I know know the cause if not the cure for my problem

**Reply by thammer67** [reply]
I have modified the `c_clientsince` table column to use DATETIME instead of DATE and re-ran `prisma generate` and restarted my Nuxt dev server but I am still getting the error when I do `SELECT c_clientsince`. Note that I get no error when I `SELECT c_dtem` which is also a DATETIME column. Do I need to update anything else? The docs do not indicate if I need to do anything else.

**Reply by janpio** [reply, marked as answer]
This might be related: https://github.com/prisma/prisma/issues/5006

**Reply by thammer67** [reply]
Indeed, janpio. That seems to be exactly what the problem is. My legacy db has zero-ized date values. Apparently I need to change that. Thank you!

On a side note, it was quite difficult and time consuming to track down this issue. Prisma could have found this issue and spit out a helpful error indicating why it was failing. But at least it's been identified and I know what to do.

**Reply by janpio** [reply]
Can you open an issue for that? I think it is an interesting suggestion to throw a more targetted error message on encountered zero dates until we can add support for them.

**Reply by nurul3101** [reply]
Thank you for opening the GitHub Issue 🙌 
https://github.com/prisma/prisma/issues/24522

Also, thanks Jan for sharing the link to the GitHub Issue causing this 🙏

**Reply by thammer67** [reply]
Hey just trying to help make Prisma better :-)
